dbos/system_database.go

```diff
@@ -139,52 +139,137 @@ const (
-	separator := "?"
-	if parsedURL.RawQuery != "" {
-		separator = "&"
+	// Create the DBOS schema if it doesn't exist
+	_, err = tx.Exec(ctx, "CREATE SCHEMA IF NOT EXISTS dbos")
+	if err != nil {
+		return fmt.Errorf("failed to create DBOS schema: %v", err)
 	}
-	databaseURL += separator + "x-migrations-table=" + _DBOS_MIGRATION_TABLE
 
-	// Create migration source from embedded files
-	d, err := iofs.New(migrationFiles, "migrations")
+	// Create the migrations table if it doesn't exist
+	createTableQuery := fmt.Sprintf(`CREATE TABLE IF NOT EXISTS dbos.%s (
+		version BIGINT NOT NULL PRIMARY KEY
+	)`, _DBOS_MIGRATION_TABLE)
+
+	_, err = tx.Exec(ctx, createTableQuery)
 	if err != nil {
-		return fmt.Errorf("failed to create migration source: %v", err)
+		return fmt.Errorf("failed to create migrations table: %v", err)
+	}
+
+	// Get current migration version
+	var currentVersion int64 = 0
+	query := fmt.Sprintf("SELECT version FROM dbos.%s LIMIT 1", _DBOS_MIGRATION_TABLE)
+	err = tx.QueryRow(ctx, query).Scan(&currentVersion)
+	if err != nil && err != pgx.ErrNoRows {
+		return fmt.Errorf("failed to get current migration version: %v", err)
 	}
 
-	// Create migrator
-	m, err := migrate.NewWithSourceInstance("iofs", d, databaseURL)
+	// Read and parse migration files
+	migrations, err := parseMigrationFiles()
 	if err != nil {
-		return fmt.Errorf("failed to create migrator: %v", err)
+		return fmt.Errorf("failed to parse migration files: %v", err)
 	}
-	defer m.Close()
 
-	// Run migrations
-	// FIXME: tolerate errors when the migration is bcz we run an older version of transact
-	if err := m.Up(); err != nil && err != migrate.ErrNoChange {
-		return fmt.Errorf("failed to run migrations: %v", err)
+	// Apply migrations starting from the next version
+	for _, migration := range migrations {
+		if migration.version <= currentVersion {
+			continue
+		}
+
+		// Execute the migration SQL
+		_, err = tx.Exec(ctx, migration.sql)
+		if err != nil {
+			return fmt.Errorf("failed to execute migration %d: %v", migration.version, err)
+		}
+
+		// Update the migration version
+		if currentVersion == 0 {
+			// Insert first migration record
+			insertQuery := fmt.Sprintf("INSERT INTO dbos.%s (version) VALUES ($1)", _DBOS_MIGRATION_TABLE)
+			_, err = tx.Exec(ctx, insertQuery, migration.version)
+		} else {
+			// Update existing migration record
+			updateQuery := fmt.Sprintf("UPDATE dbos.%s SET version = $1", _DBOS_MIGRATION_TABLE)
+			_, err = tx.Exec(ctx, updateQuery, migration.version)
+		}
+		if err != nil {
+			return fmt.Errorf("failed to update migration version to %d: %v", migration.version, err)
+		}
+
+		currentVersion = migration.version
+	}
+
+	// Commit the transaction
+	if err := tx.Commit(ctx); err != nil {
+		return fmt.Errorf("failed to commit migration transaction: %v", err)
 	}
 
 	return nil
 }
 
+type migrationFile struct {
+	version int64
+	sql     string
+}
+
+func parseMigrationFiles() ([]migrationFile, error) {
```

Comment:
Let's not do any of this file parsing stuff. How does this even work in a compiled binary? Just make them an array in memory.

Reply:
Let's just do strings. Way simpler.

Reply:
I don't think its simpler. sql files are much more manageable. We can embed them as strings, one by one, in the file if we need, so we don't have to do the sorting/"parsing", but having them in the files is really bad imo. We have to do it in the other languages because they are not compiled, but here we have the facility to do this both correctly and manageably

Reply:
I'm looking at this code and I think strings are simpler. You don't have to worry about parsing file names or handle errors if the migrations are just an array of strings. You can literally delete this entire function.